core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -596,6 +591,13 @@ class SparkContext(config: SparkConf) extends Logging {
       _conf.set(APP_ATTEMPT_ID, attemptId)
       _env.blockManager.blockStoreClient.setAppAttemptId(attemptId)
     }
+
+    // initialize after application id and attempt id has been initialized
```

Comment:
Shuffle driver components typically depend on the application/attempt id for managing their metadata.
Initializing it after these values are set in `SparkConcext` allows the `ShuffleDriverComponents` to rely on this - instead of using hacks like `SparkListener` waiting for application start, etc for the actual initialization.